pydantic_ai_graph/pydantic_ai_graph/nodes.py

```diff
@@ -1,113 +1,93 @@
 from __future__ import annotations as _annotations
 
-from abc import ABC, ABCMeta, abstractmethod
+from abc import abstractmethod
 from dataclasses import dataclass
 from functools import cache
-from typing import Any, ClassVar, Generic, get_args, get_origin, get_type_hints
+from typing import Any, ClassVar, Generic, get_origin, get_type_hints
 
-from typing_extensions import TypeVar
+from typing_extensions import Never, TypeVar
 
 from . import _utils
 from .state import StateT
 
-__all__ = (
-    'NodeInputT',
-    'GraphOutputT',
-    'GraphContext',
-    'End',
-    'BaseNode',
-    'NodeDef',
-)
+__all__ = ('GraphContext', 'End', 'BaseNode', 'NodeDef')
 
-NodeInputT = TypeVar('NodeInputT', default=Any)
-GraphOutputT = TypeVar('GraphOutputT', default=Any)
+RunEndT = TypeVar('RunEndT', default=None)
```

Comment:
I think this should not have a default.

Reply:
If we get rid of the default, then it needs to come before StateT everywhere. I'm okay with that, but just pointing it out. I will make that change

Reply:
Actually, it's awkward because, since parameters with defaults need to go after parameters without defaults, unless we drop the default value of `StateT` for `Graph`, we'd have to put `RunEndT` before `StateT` on `Graph`. But then you have to make a decision of whether you want consistency in parameter ordering on `BaseNode`, or convenience when omitting default parameters:

* If we keep the ordering of parameters on `BaseNode` as it currently is, then `Graph` has the parameters in one order and `BaseNode` has it in the other, which feels like a recipe for confusion.
* If we change the ordering of parameters on `BaseNode`, then to explicitly specify the `StateT` type you need to specify the `NodeRunEndT`, but explicitly specifying the `StateT` will be a lot more common than explicitly specifying the node run end type, so that feels pretty unfortunate.

Given the above, I now feel like we should _either_:
* keep a default of `None` for the `RunEndT` on `Graph`, or
* drop the default of `None` for the `StateT` on `Graph`.

I'm okay with either, lmk what you think. (I'm also okay with the first two bullets' suggestions, if you really prefer, but I think they are worse than making it so the graph typevars either both have defaults or both don't, and preserving parameter ordering).

Reply:
I'll note that trying to do the refactor, I think things get a little more complicated if we remove the default, and I'm not sure that having the default of `None` is that crazy given that it provides a way to end the run without returning a meaningful value, which may be more common than you'd expect if it ends up being common to create agents that do mutating tool calls before finishing rather than returning values in the end.